src/propmcwf.jl

```diff
@@ -0,0 +1,134 @@
+    rho::QA
+    ntraj::Int
+    decomp
+    QuMCWFEnsemble(rho, ntraj, decomp) = new(rho, ntraj, decomp)
+end
+
+QuMCWFEnsemble{QM<:QuBase.AbstractQuMatrix}(rho::QM, ntraj=500) = QuMCWFEnsemble{QM}(rho, ntraj, eigfact(full(coeffs(rho))))
+QuMCWFEnsemble{QV<:QuBase.AbstractQuVector}(psi::QV, ntraj=500) = QuMCWFEnsemble{QV}(psi, ntraj, nothing)
+
+Base.length(en::QuMCWFEnsemble) = en.ntraj
+
+@doc """
+Input Parameters : QuMCWFEnsemble
+
+Returns the starting iterator state of the Ensemble
+""" ->
+Base.start(mcwfensemble::QuMCWFEnsemble) = 1
+
+@doc """
+Input Parameters : QuMCWFEnsemble and iterator state
+
+Returns the next state by using `draw` function
+to draw the next state and iterator state.
+""" ->
+function Base.next(mcwfensemble::QuMCWFEnsemble, i::Int)
+    state = draw(mcwfensemble)
+    return (state, i+1)
+end
+
+@doc """
+Input Parameters : QuMCWFEnsemble and iterator state
+
+Compares the iterator state with the number of trajectories and
+returns true or false depending on the comparison.
+""" ->
+function Base.done(mcwfensemble::QuMCWFEnsemble, i::Int)
+    i > mcwfensemble.ntraj
+end
+
+@doc """
+Input Parameters : QuMCWFEnsemble which takes a AbstractQuMatrix as a parameter
+
+Returns state vector using the decomposition of AbstractQuMatrix.
+""" ->
+function draw{QM<:QuBase.AbstractQuMatrix}(mcwfensemble::QuMCWFEnsemble{QM})
+    r = rand() # draw random number from [0,1)
+    pacc = 0.
+    for i=1:length(mcwfensemble.decomp[:values])
+        pacc = pacc + mcwfensemble.decomp[:values][i]
+        if pacc >= r
+            return QuArray(complex(mcwfensemble.decomp[:vectors][:,i]))
+        end
+    end
+end
+
+@doc """
+Input Parameters : QuMCWFEnsemble which takes a AbstractQuVector as a parameter
+
+Returns a copy of the state passed through the input parameter.
+""" ->
+function draw{QV<:QuBase.AbstractQuVector}(mcwfensemble::QuMCWFEnsemble{QV})
+    return copy(mcwfensemble.rho)
+end
+
+function propagate(prob::QuMCWF, eq::QuLindbladMasterEq, t, current_t, psi)
+    const jtol = 1.e-6  # jump tolerance
+    # get information of QME
+    heff = eff_hamiltonian(eq)
+    c_ops = eq.collapse_ops
+    dt = t - current_t
+    accdt = 0.
+    while accdt < dt
+        # propagate one time-step
+        # the following only works for time independent cases (WIP for time dependent)
+        psi1 = QuDynamics.propagate(get(prob.options, :solver, QuExpmV()), QuSchrodingerEq(heff), current_t + dt, current_t, psi)
+        if norm(psi1)^2 > prob.eps + jtol/2
+            psi = copy(psi1)  # no jump
+            accdt = accdt + dt
+            dt = (t - current_t) - accdt
+        elseif norm(psi1)^2 < prob.eps - jtol/2
+            dt = dt/2		# jump in the current interval
+        else
+            if isempty(c_ops)
+                psi = copy(psi1)
+            else
+                PnS = 0.
+                for cind in c_ops
+                    PnS = PnS + real(expectation(psi1, cind'*cind))
```

Comment:
It is a bit problematic to use `expectation` since it is not yet in `QuBase`. You could replace it by `dot(psi1, cind'*cind*psi1)` and put a comment noting that it has to be replaced eventually.